lib/sidekiq/job/iterable.rb

```diff
@@ -0,0 +1,245 @@
+              arguments.pop["sidekiq_iteration"]
+            else
+              {}
+            end
+
+          @executions = options["executions"] || 0
+          @cursor_position = options["cursor_position"]
+          @times_interrupted = options["times_interrupted"] || 0
+          @total_time = options["total_time"] || 0
+        end
+
+        def interruptible_perform(*arguments)
+          @executions += 1
+          @start_time = Time.now.utc
+
+          enumerator = build_enumerator(*arguments, cursor: cursor_position)
+          unless enumerator
+            logger.debug("`build_enumerator` returned nil. Skipping the job.")
+            return
+          end
+
+          assert_enumerator!(enumerator)
+
+          if executions == 1 && times_interrupted == 0
+            on_start
+          else
+            on_resume
+          end
+
+          completed = catch(:abort) do
+            iterate_with_enumerator(enumerator, arguments)
+          end
+
+          on_shutdown
+          completed = handle_completed(completed)
+
+          if @needs_reenqueue
+            reenqueue_iteration_job
+          elsif completed
+            on_complete
+
+            logger.info(
+              format("Completed iterating. times_interrupted=%d total_time=%.3f", times_interrupted, total_time)
+            )
+          end
+        end
+
+        def iterate_with_enumerator(enumerator, arguments)
+          found_record = false
+          @needs_reenqueue = false
+
+          enumerator.each do |object_from_enumerator, index|
+            found_record = true
+            around_iteration do
+              each_iteration(object_from_enumerator, *arguments)
+            end
+            @cursor_position = index
+
+            if job_should_exit?
+              @needs_reenqueue = true
+              return false
+            end
+          end
+
+          unless found_record
+            logger.debug {
+              "Enumerator found nothing to iterate! " \
+              "times_interrupted=#{times_interrupted} cursor_position=#{cursor_position}"
+            }
+          end
+
+          true
+        ensure
+          adjust_total_time
+        end
+
+        def reenqueue_iteration_job
+          logger.debug { "Interrupting and re-enqueueing the job (cursor_position=#{cursor_position})" }
+
+          @times_interrupted += 1
+
+          arguments = @arguments
+          arguments.push(
+            "iteration" => {
+              "executions" => executions,
+              "cursor_position" => cursor_position,
+              "times_interrupted" => times_interrupted,
+              "total_time" => total_time
+            }
+          )
+
+          retry_backoff = self.class.get_sidekiq_options.dig("iteration", "retry_backoff") ||
+                          Sidekiq.default_configuration["retry_backoff"]
+          self.class.perform_in(retry_backoff, *arguments)
```

Comment:
Is there a better way to reschedule itself than keeping arguments in the job and then explicitly calling `perform_in`?

For example, there is a helper in the ActiveJob for the similar purpose - https://api.rubyonrails.org/classes/ActiveJob/Exceptions.html#method-i-retry_job. Maybe we need something similar?

Reply:
Good question, it might be time to consider a API for this purpose. Generally I don't like the idea of a job performing its own "meta-logic" around scheduling because it makes testing difficult. It's a massive violation of the Single Responsibility Principle. But this API wouldn't be used by the job itself, but rather your iteration logic.